array_api_compat/common/_helpers.py

```diff
@@ -801,6 +801,174 @@ def size(x):
         return None
     return math.prod(x.shape)
 
+def is_writeable_array(x):
+    """
+    Return False if x.__setitem__ is expected to raise; True otherwise
+    """
+    if is_numpy_array(x):
+        return x.flags.writeable
+    if is_jax_array(x):
+        return False
+    return True
+
+_undef = object()
+
+def at(x, idx=_undef, /):
+    """
+    Update operations for read-only arrays.
+
+    This implements ``jax.numpy.ndarray.at`` for all backends.
+    Writeable arrays may be updated in place; you should not rely on it.
+
+    Keyword arguments (e.g. ``indices_are_sorted``) are passed to JAX and are
+    quietly ignored for backends that don't support them.
+
+    Examples
+    --------
+    Given either of these equivalent expressions::
+
+      x = at(x)[1].add(2)
+      x = at(x, 1).add(2)
+
+    If x is a JAX array, they are the same as::
+
+      x = x.at[1].add(x)
+
+    If x is a read-only numpy array, they are the same as::
+
+      x = x.copy()
+      x[1] += 2
+
+    Otherwise, they are the same as::
+
+      x[1] += 2
+
+    Warning
+    -------
+    You should always immediately overwrite the parameter array::
+
+        x = at(x, 0).set(2)
+
+    The anti-pattern below must be avoided, as it will result in different behaviour
+    on read-only versus writeable arrays:
+
+        x = xp.asarray([0, 0, 0])
+        y = at(x, 0).set(2)
+        z = at(x, 1).set(3)
+
+    In the above example, y == [2, 0, 0] and z == [0, 3, 0] when x is read-only,
+    whereas y == z == [2, 3, 0] when x is writeable!
+
+    See Also
+    --------
+    https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html
+    """
+    if is_jax_array(x):
+        return x.at
+    if is_numpy_array(x) and not x.flags.writeable:
+        x = x.copy()
```

Comment:
perhaps we would want an optional kwarg to error instead of copy here? I suppose the semantics for `copy=True/None/False` used elsewhere in the standard could fit?

Reply:
This is a great idea